cardano: Implement SLIP-0023 and add SLIP-0039 support for Cardano.

andrewkozlik opened the pull request:
This should fix https://github.com/trezor/trezor-firmware/issues/272

I created SLIP-0023 to solve this issue, see https://github.com/satoshilabs/slips/blob/andrewkozlik/slip-0023/slip-0023.md. In https://forum.cardano.org/t/cardano-with-slip-0039/25443 I asked people to make comments on the new specification in the pull request https://github.com/satoshilabs/slips/pull/712.

- I have not checked the code, so please review carefully.
- I have not tested this.
- I have not been able to create test vectors yet. If anybody knows about a tool like https://iancoleman.io/bip39/ for Cardano, where I can input a private key and chain code and derive an address, then please let me know.
- There is a bizzare merge conflict, that I am now too tired to resolve - mnemonic.get() now returns module instead of type, WTF?

tsusanka commented:
I've rebased on master and fixed the conflicts. It's currently in tsusanka/272-rebased and I'll force push here after I've confirmed tomorrow that I've done it correctly.

andrewkozlik commented:
> I've rebased on master and fixed the conflicts. It's currently in tsusanka/272-rebased and I'll force push here tomorrow after I confirm I've done it correctly. Or check it out, if you have a sec.

The rebase looks fine except for mnemonic.get_seed(). Something happened with _stop_progress() at the end of the function. I only added the "mnemonic_secret is None" check at the beginning because of mypy warnings, I didn't touch the rest of the function, so I don't know what this is about.

SebastienGllmt commented:
Unfortunately there is no website like to easily play around with Cardano's BIP-39 scheme. The best that we have right now is that you can use the Cardano [WASM bindings](https://www.npmjs.com/package/cardano-wallet-browser) on a REPL website to easily play with it and share links. For example, here is a link to generate addresses from a mnemonic https://repl.it/repls/IndolentWarmheartedDehardwarization

andrewkozlik commented:
Thanks @SebastienGllmt, that's what I was looking for.

@tsusanka I will generate some test vectors this evening.

andrewkozlik commented:
- I rebased this branch on top of master. (In tsusanka/272-rebased the function mnemonic.get_seed() is indeed merged incorrectly.)
- I created some test vectors and added them to core/tests. I will add them to the SLIP-0023 spec shortly.

andrewkozlik commented:
Tested on a SLIP-0039 initialized device and checked that it gives the same Cardano addresses as are in the test vectors. So this PR should be good to go.
